packages/flutter/lib/src/widgets/routes.dart

```diff
@@ -1933,12 +2011,34 @@ abstract class ModalRoute<T> extends TransitionRoute<T> with LocalHistoryRoute<T
 
   @override
   void didChangeNext(Route<dynamic>? nextRoute) {
+    // In some cases, we need to get the delegated transition from a route before
+    // it is built, but some routes need to derive a delegated transition from
+    // the context. So if the next route does not have a `BuildContext` or a
+    // delegatedTransition, call nextRoute.setDelegatedTransition using the
+    // current `BuildContext`.
+    if (nextRoute is ModalRoute<T> && nextRoute.delegatedTransition == null && nextRoute.subtreeContext == null && this.subtreeContext != null) {
+      nextRoute.setDelegatedTransition(this.subtreeContext!);
+    }
+    if (nextRoute is ModalRoute<T> && canTransitionTo(nextRoute) && nextRoute.delegatedTransition.runtimeType != this.delegatedTransition.runtimeType) {
+      if (receivedTransition.runtimeType != nextRoute.delegatedTransition.runtimeType) {
+        receivedTransition = nextRoute.delegatedTransition;
+      }
+    } else {
+      receivedTransition = null;
+    }
     super.didChangeNext(nextRoute);
     changedInternalState();
   }
 
   @override
   void didPopNext(Route<dynamic> nextRoute) {
+    if (nextRoute is ModalRoute<T> && canTransitionTo(nextRoute) && nextRoute.delegatedTransition.runtimeType != this.delegatedTransition.runtimeType) {
+      if (receivedTransition.runtimeType != nextRoute.delegatedTransition.runtimeType) {
```

Comment:
even if they have the same runtimeType, they can still be different. Why we skip updating the `receivedTransition` in this case?

Reply:
Because we have at least one case where we intentionally want to not get a `receivedTransition` in that case. For example, the Zoom transition can have differing rasterization depending on the page. So if the current page is not snapshotted, but the next page is, we want to not override the current page with a snapshotted transition. Having it go by the type lets us do that decently smoothly.